From a97644af5500b9ad3bcb57563661c67f9ce6f5f1 Mon Sep 17 00:00:00 2001
From: Alexandru Croitor <alexandru.croitor@qt.io>
Date: Thu, 4 Sep 2025 14:03:16 +0200
Subject: [PATCH 120/553] CMake: Rename TYPE to SBOM_ENTITY_TYPE in SBOM API

Introduce two new options for SBOM entity type specification:
- SBOM_ENTITY_TYPE: has the highest priority
- DEFAULT_SBOM_ENTITY_TYPE: Used as a fallback when no explicit type
  is given.

Deprecate the old pre-existing TYPE option in favor of either
SBOM_ENTITY_TYPE or DEFAULT_SBOM_ENTITY_TYPE.

The reason is to allow forwarding these option names directly via
functions like qt_internal_add_sbom, or the Creator equivalents,
without having a potential clash with a very generic name like TYPE.

Adjust the code to use the new options inside the various
implementation functions, but keep the old TYPE option for
compatibility in the public facing APIs, to not break existing
callers.

Pick-to: 6.9 6.8
Task-number: QTBUG-134894
Change-Id: I36cd6c4a5dab19221efacbe32688eab7db4f142f
Reviewed-by: Alexey Edelev <alexey.edelev@qt.io>
(cherry picked from commit 928f6cd209c42441c62e7a92e28f2e1e99b02c7e)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 cmake/Qt3rdPartyLibraryHelpers.cmake          |   2 +-
 cmake/QtAppHelpers.cmake                      |   2 +-
 cmake/QtFindPackageHelpers.cmake              |   2 +-
 cmake/QtModuleHelpers.cmake                   |   4 +-
 cmake/QtPlatformTargetHelpers.cmake           |   2 +-
 cmake/QtPluginHelpers.cmake                   |   2 +-
 cmake/QtPublicDependencyHelpers.cmake         |   2 +-
 cmake/QtPublicSbomAttributionHelpers.cmake    |  16 ++-
 cmake/QtPublicSbomFileHelpers.cmake           |  29 ++---
 cmake/QtPublicSbomHelpers.cmake               | 111 ++++++++++++------
 cmake/QtPublicSbomQtEntityHelpers.cmake       |  54 ++++++---
 cmake/QtPublicSbomSystemDepHelpers.cmake      |  11 +-
 cmake/QtTargetHelpers.cmake                   |   2 +-
 cmake/QtToolHelpers.cmake                     |   2 +-
 qmake/CMakeLists.txt                          |   2 +-
 src/tools/bootstrap/CMakeLists.txt            |   2 +-
 .../cmake/RunCMake/Sbom/common_targets.cmake  |  24 ++++
 17 files changed, 181 insertions(+), 88 deletions(-)

diff --git a/cmake/Qt3rdPartyLibraryHelpers.cmake b/cmake/Qt3rdPartyLibraryHelpers.cmake
index 9c86e26076d..d6ccaa0087d 100644
--- a/cmake/Qt3rdPartyLibraryHelpers.cmake
+++ b/cmake/Qt3rdPartyLibraryHelpers.cmake
@@ -343,7 +343,7 @@ function(qt_internal_add_3rdparty_library target)
 
     if(QT_GENERATE_SBOM)
         set(sbom_args "")
-        list(APPEND sbom_args TYPE QT_THIRD_PARTY_MODULE)
+        list(APPEND sbom_args DEFAULT_SBOM_ENTITY_TYPE QT_THIRD_PARTY_MODULE)
 
         if(NOT will_install)
             list(APPEND sbom_args NO_INSTALL)
diff --git a/cmake/QtAppHelpers.cmake b/cmake/QtAppHelpers.cmake
index 00613275416..895788c3a25 100644
--- a/cmake/QtAppHelpers.cmake
+++ b/cmake/QtAppHelpers.cmake
@@ -122,7 +122,7 @@ function(qt_internal_add_app target)
 
     if(QT_GENERATE_SBOM)
         set(sbom_args "")
-        list(APPEND sbom_args TYPE QT_APP)
+        list(APPEND sbom_args DEFAULT_SBOM_ENTITY_TYPE QT_APP)
 
         qt_get_cmake_configurations(cmake_configs)
         foreach(cmake_config IN LISTS cmake_configs)
diff --git a/cmake/QtFindPackageHelpers.cmake b/cmake/QtFindPackageHelpers.cmake
index aadf0bf1f7f..29c12dcd826 100644
--- a/cmake/QtFindPackageHelpers.cmake
+++ b/cmake/QtFindPackageHelpers.cmake
@@ -234,7 +234,7 @@ macro(qt_find_package)
                 if(NOT "${ARGV0}" STREQUAL "Qt6")
                     _qt_internal_sbom_record_system_library_usage(
                         "${qt_find_package_target_name}"
-                        TYPE SYSTEM_LIBRARY
+                        SBOM_ENTITY_TYPE SYSTEM_LIBRARY
                         FRIENDLY_PACKAGE_NAME "${ARGV0}"
                         ${_qt_find_package_sbom_args}
                     )
diff --git a/cmake/QtModuleHelpers.cmake b/cmake/QtModuleHelpers.cmake
index 75c152a0a77..dcbb85238c3 100644
--- a/cmake/QtModuleHelpers.cmake
+++ b/cmake/QtModuleHelpers.cmake
@@ -1024,9 +1024,9 @@ set(QT_ALLOW_MISSING_TOOLS_PACKAGES TRUE)")
 
         # 3rd party header modules should not be treated as Qt modules.
         if(arg_IS_QT_3RD_PARTY_HEADER_MODULE)
-            list(APPEND sbom_args TYPE QT_THIRD_PARTY_MODULE)
+            list(APPEND sbom_args DEFAULT_SBOM_ENTITY_TYPE QT_THIRD_PARTY_MODULE)
         else()
-            list(APPEND sbom_args TYPE QT_MODULE)
+            list(APPEND sbom_args DEFAULT_SBOM_ENTITY_TYPE QT_MODULE)
         endif()
 
         qt_get_cmake_configurations(configs)
diff --git a/cmake/QtPlatformTargetHelpers.cmake b/cmake/QtPlatformTargetHelpers.cmake
index dba39721b69..4125da5c999 100644
--- a/cmake/QtPlatformTargetHelpers.cmake
+++ b/cmake/QtPlatformTargetHelpers.cmake
@@ -66,7 +66,7 @@ function(qt_internal_setup_public_platform_target)
     qt_internal_generate_pkg_config_file(Platform)
 
     qt_internal_add_sbom(Platform
-        TYPE QT_MODULE
+        SBOM_ENTITY_TYPE QT_MODULE
         ATTRIBUTION_FILE_DIR_PATHS
             "${PROJECT_SOURCE_DIR}/cmake/3rdparty/extra-cmake-modules"
             "${PROJECT_SOURCE_DIR}/cmake/3rdparty/kwin"
diff --git a/cmake/QtPluginHelpers.cmake b/cmake/QtPluginHelpers.cmake
index 00114bf0306..28a2bc36cc6 100644
--- a/cmake/QtPluginHelpers.cmake
+++ b/cmake/QtPluginHelpers.cmake
@@ -456,7 +456,7 @@ endif()"
 
     if(QT_GENERATE_SBOM)
         set(sbom_args "")
-        list(APPEND sbom_args TYPE QT_PLUGIN)
+        list(APPEND sbom_args DEFAULT_SBOM_ENTITY_TYPE QT_PLUGIN)
 
         qt_get_cmake_configurations(configs)
         foreach(config IN LISTS configs)
diff --git a/cmake/QtPublicDependencyHelpers.cmake b/cmake/QtPublicDependencyHelpers.cmake
index 951781c3d33..b4a0342ad87 100644
--- a/cmake/QtPublicDependencyHelpers.cmake
+++ b/cmake/QtPublicDependencyHelpers.cmake
@@ -64,7 +64,7 @@ macro(_qt_internal_find_third_party_dependencies target target_dep_list)
 
                 _qt_internal_sbom_record_system_library_usage(
                     "${__qt_${target}_provided_target}"
-                    TYPE SYSTEM_LIBRARY
+                    SBOM_ENTITY_TYPE SYSTEM_LIBRARY
                     FRIENDLY_PACKAGE_NAME "${__qt_${target}_pkg}"
                     ${__qt_${target}_sbom_args}
                 )
diff --git a/cmake/QtPublicSbomAttributionHelpers.cmake b/cmake/QtPublicSbomAttributionHelpers.cmake
index 2d25fcd0b39..8b114d8eee7 100644
--- a/cmake/QtPublicSbomAttributionHelpers.cmake
+++ b/cmake/QtPublicSbomAttributionHelpers.cmake
@@ -58,9 +58,11 @@ function(_qt_internal_sbom_handle_qt_attribution_files out_prefix_outer)
     # In which case we don't want to proagate options like CPE to the child attribution targets,
     # because the CPE is meant for the parent target.
     set(propagate_sbom_options_to_new_attribution_targets TRUE)
-    if(arg___QT_INTERNAL_HANDLE_QT_ENTITY_ATTRIBUTION_FILES)
-        _qt_internal_sbom_is_qt_entity_type("${arg_TYPE}" is_qt_entity_type)
-        if(is_qt_entity_type)
+    if(arg___QT_INTERNAL_HANDLE_QT_ENTITY_ATTRIBUTION_FILES
+            AND arg_ATTRIBUTION_PARENT_TARGET_SBOM_ENTITY_TYPE)
+        _qt_internal_sbom_is_qt_entity_type("${arg_ATTRIBUTION_PARENT_TARGET_SBOM_ENTITY_TYPE}"
+            parent_is_qt_entity_type)
+        if(parent_is_qt_entity_type)
             set(arg_CREATE_SBOM_FOR_EACH_ATTRIBUTION TRUE)
         endif()
     endif()
@@ -240,6 +242,12 @@ function(_qt_internal_sbom_handle_qt_attribution_files out_prefix_outer)
                     # file(GENERATE) multiple times with the same file name, but different content.
                     list(REMOVE_ITEM sbom_single_args FRIENDLY_PACKAGE_NAME)
 
+                    # Filter out the sbom entity types, they are specified manually.
+                    list(REMOVE_ITEM sbom_single_args
+                        SBOM_ENTITY_TYPE
+                        DEFAULT_SBOM_ENTITY_TYPE
+                    )
+
                     _qt_internal_forward_function_args(
                         FORWARD_APPEND
                         FORWARD_PREFIX arg
@@ -264,7 +272,7 @@ function(_qt_internal_sbom_handle_qt_attribution_files out_prefix_outer)
 
                 _qt_internal_add_sbom("${attribution_target}"
                     IMMEDIATE_FINALIZATION
-                    TYPE "${attribution_entity_type}"
+                    SBOM_ENTITY_TYPE "${attribution_entity_type}"
                     ATTRIBUTION_FILE_PATHS "${attribution_file_path}"
                     ATTRIBUTION_ENTRY_INDEX "${entry_index}"
                     NO_CURRENT_DIR_ATTRIBUTION
diff --git a/cmake/QtPublicSbomFileHelpers.cmake b/cmake/QtPublicSbomFileHelpers.cmake
index 2c3c9d7ff8f..6fca87b04ac 100644
--- a/cmake/QtPublicSbomFileHelpers.cmake
+++ b/cmake/QtPublicSbomFileHelpers.cmake
@@ -9,7 +9,7 @@ function(_qt_internal_sbom_handle_target_binary_files target)
         FRAMEWORK
     )
     set(single_args
-        TYPE
+        SBOM_ENTITY_TYPE
         SPDX_ID
         LICENSE_EXPRESSION
         INSTALL_PREFIX
@@ -60,9 +60,10 @@ function(_qt_internal_sbom_handle_target_binary_files target)
         CUSTOM_NO_INFIX
     )
 
-    if(NOT arg_TYPE IN_LIST supported_types)
+    if(NOT arg_SBOM_ENTITY_TYPE IN_LIST supported_types)
         message(FATAL_ERROR
-            "Unsupported target TYPE '${arg_TYPE}' for target '${target}' during SBOM creation.")
+            "Unsupported target TYPE '${arg_SBOM_ENTITY_TYPE}' for target '${target}' during "
+            "SBOM creation.")
     endif()
 
     set(types_without_binary_files
@@ -82,9 +83,9 @@ function(_qt_internal_sbom_handle_target_binary_files target)
 
     get_target_property(target_type ${target} TYPE)
 
-    if(arg_TYPE IN_LIST types_without_binary_files)
+    if(arg_SBOM_ENTITY_TYPE IN_LIST types_without_binary_files)
         message(DEBUG "Target ${target} has no binary files to reference in the SBOM "
-            "because it has the ${arg_TYPE} type.")
+            "because it has the ${arg_SBOM_ENTITY_TYPE} type.")
         return()
     endif()
 
@@ -109,7 +110,7 @@ function(_qt_internal_sbom_handle_target_binary_files target)
     set(file_common_options "")
 
     list(APPEND file_common_options PACKAGE_SPDX_ID "${package_spdx_id}")
-    list(APPEND file_common_options PACKAGE_TYPE "${arg_TYPE}")
+    list(APPEND file_common_options PACKAGE_TYPE "${arg_SBOM_ENTITY_TYPE}")
 
     if(arg_COPYRIGHTS)
         list(APPEND file_common_options COPYRIGHTS "${arg_COPYRIGHTS}")
@@ -131,9 +132,9 @@ function(_qt_internal_sbom_handle_target_binary_files target)
         set(library_path_kind LIBRARY_PATH)
     endif()
 
-    if(arg_TYPE STREQUAL "QT_TOOL"
-            OR arg_TYPE STREQUAL "QT_APP"
-            OR arg_TYPE STREQUAL "EXECUTABLE")
+    if(arg_SBOM_ENTITY_TYPE STREQUAL "QT_TOOL"
+            OR arg_SBOM_ENTITY_TYPE STREQUAL "QT_APP"
+            OR arg_SBOM_ENTITY_TYPE STREQUAL "EXECUTABLE")
 
         set(valid_executable_types
             "EXECUTABLE"
@@ -156,7 +157,7 @@ function(_qt_internal_sbom_handle_target_binary_files target)
             PATH_SUFFIX "${path_suffix}"
             OPTIONS ${file_common_options}
         )
-    elseif(arg_TYPE STREQUAL "QT_PLUGIN")
+    elseif(arg_SBOM_ENTITY_TYPE STREQUAL "QT_PLUGIN")
         if(NOT (target_type STREQUAL "SHARED_LIBRARY"
                 OR target_type STREQUAL "STATIC_LIBRARY"
                 OR target_type STREQUAL "MODULE_LIBRARY"))
@@ -168,10 +169,10 @@ function(_qt_internal_sbom_handle_target_binary_files target)
             PATH_SUFFIX "${path_suffix}"
             OPTIONS ${file_common_options}
         )
-    elseif(arg_TYPE STREQUAL "QT_MODULE"
-            OR arg_TYPE STREQUAL "QT_THIRD_PARTY_MODULE"
-            OR arg_TYPE STREQUAL "LIBRARY"
-            OR arg_TYPE STREQUAL "THIRD_PARTY_LIBRARY_WITH_FILES"
+    elseif(arg_SBOM_ENTITY_TYPE STREQUAL "QT_MODULE"
+            OR arg_SBOM_ENTITY_TYPE STREQUAL "QT_THIRD_PARTY_MODULE"
+            OR arg_SBOM_ENTITY_TYPE STREQUAL "LIBRARY"
+            OR arg_SBOM_ENTITY_TYPE STREQUAL "THIRD_PARTY_LIBRARY_WITH_FILES"
         )
         if(WIN32 AND target_type STREQUAL "SHARED_LIBRARY")
             _qt_internal_sbom_handle_multi_config_target_binary_file(${target}
diff --git a/cmake/QtPublicSbomHelpers.cmake b/cmake/QtPublicSbomHelpers.cmake
index 144d462acea..3935b917609 100644
--- a/cmake/QtPublicSbomHelpers.cmake
+++ b/cmake/QtPublicSbomHelpers.cmake
@@ -611,7 +611,7 @@ macro(_qt_internal_get_sbom_purl_handling_options opt_args single_args multi_arg
     set(${opt_args} "")
     set(${single_args}
         SUPPLIER
-        TYPE
+        SBOM_ENTITY_TYPE
         PACKAGE_VERSION
     )
     set(${multi_args} "")
@@ -649,6 +649,8 @@ macro(_qt_internal_get_sbom_add_target_common_options opt_args single_args multi
         __QT_INTERNAL_HANDLE_QT_ENTITY_ATTRIBUTION_FILES
     )
     set(${single_args}
+        DEFAULT_SBOM_ENTITY_TYPE
+        SBOM_ENTITY_TYPE
         PACKAGE_VERSION
         FRIENDLY_PACKAGE_NAME
         SUPPLIER
@@ -659,6 +661,7 @@ macro(_qt_internal_get_sbom_add_target_common_options opt_args single_args multi
         DOWNLOAD_LOCATION
         ATTRIBUTION_ENTRY_INDEX
         ATTRIBUTION_PARENT_TARGET
+        ATTRIBUTION_PARENT_TARGET_SBOM_ENTITY_TYPE
         SBOM_PACKAGE_COMMENT
     )
     set(${multi_args}
@@ -702,7 +705,7 @@ macro(_qt_internal_get_sbom_add_target_options opt_args single_args multi_args)
         NO_INSTALL
     )
     set(${single_args}
-        TYPE
+        TYPE # deprecated, use SBOM_ENTITY_TYPE or DEFAULT_SBOM_ENTITY_TYPE instead
     )
     set(${multi_args}
         LIBRARIES
@@ -716,6 +719,38 @@ macro(_qt_internal_get_sbom_add_target_options opt_args single_args multi_args)
     list(APPEND ${multi_args} ${specific_multi_args})
 endmacro()
 
+# Chooses between SBOM_ENTITY_TYPE, TYPE (deprecated) or DEFAULT_SBOM_ENTITY_TYPE and assigns it to
+# out_var.
+function(_qt_internal_map_sbom_entity_type out_var)
+    set(opt_args "")
+    set(single_args
+        TYPE # deprecated
+        SBOM_ENTITY_TYPE
+        DEFAULT_SBOM_ENTITY_TYPE
+    )
+    set(multi_args "")
+    cmake_parse_arguments(PARSE_ARGV 1 arg "${opt_args}" "${single_args}" "${multi_args}")
+
+    # Previously only TYPE existed, now there is also SBOM_ENTITY_TYPE and DEFAULT_SBOM_ENTITY_TYPE.
+    # DEFAULT_SBOM_ENTITY_TYPE is the fallback value, TYPE is deprecated but left for compatibility
+    # and SBOM_ENTITY_TYPE has the highest priority.
+    if(arg_SBOM_ENTITY_TYPE)
+        set(sbom_entity_type "${arg_SBOM_ENTITY_TYPE}")
+    elseif(arg_TYPE)
+        # deprecated code path
+        if(QT_WARN_DEPRECATED_SBOM_TYPE_OPTION)
+            message(DEPRECATION "TYPE option is deprecated, use SBOM_ENTITY_TYPE instead.")
+        endif()
+        set(sbom_entity_type "${arg_TYPE}")
+    elseif(arg_DEFAULT_SBOM_ENTITY_TYPE)
+        set(sbom_entity_type "${arg_DEFAULT_SBOM_ENTITY_TYPE}")
+    else()
+        set(sbom_entity_type "")
+    endif()
+
+    set(${out_var} "${sbom_entity_type}" PARENT_SCOPE)
+endfunction()
+
 # Generate sbom information for a given target.
 # Creates:
 # - a SPDX package for the target
@@ -738,10 +773,12 @@ function(_qt_internal_sbom_add_target target)
 
     get_target_property(target_type ${target} TYPE)
 
+    _qt_internal_map_sbom_entity_type(sbom_entity_type ${ARGN})
+
     # Mark the target as a Qt module for sbom processing purposes.
     # Needed for non-standard targets like Bootstrap and QtLibraryInfo, that don't have a Qt::
     # namespace prefix.
-    if(arg_TYPE STREQUAL QT_MODULE)
+    if(sbom_entity_type STREQUAL QT_MODULE)
         set_target_properties(${target} PROPERTIES _qt_sbom_is_qt_module TRUE)
     endif()
 
@@ -780,7 +817,7 @@ function(_qt_internal_sbom_add_target target)
     # Record the target spdx id right now, so we can refer to it in later attribution targets
     # if needed.
     _qt_internal_sbom_record_target_spdx_id(${target}
-        TYPE "${arg_TYPE}"
+        SBOM_ENTITY_TYPE "${sbom_entity_type}"
         PACKAGE_NAME "${package_name_for_spdx_id}"
         OUT_VAR package_spdx_id
     )
@@ -788,6 +825,7 @@ function(_qt_internal_sbom_add_target target)
     if(arg_USE_ATTRIBUTION_FILES)
         set(attribution_args
             ATTRIBUTION_PARENT_TARGET "${target}"
+            ATTRIBUTION_PARENT_TARGET_SBOM_ENTITY_TYPE "${sbom_entity_type}"
         )
 
         # Forward the sbom specific options when handling attribution files because those might
@@ -825,7 +863,7 @@ function(_qt_internal_sbom_add_target target)
     endif()
 
     if(arg___QT_INTERNAL_HANDLE_QT_ENTITY_TYPE_LICENSE)
-        _qt_internal_sbom_forward_sbom_add_target_options(sbom_add_target_args)
+        _qt_internal_sbom_forward_sbom_add_target_options_modified(sbom_add_target_args)
         _qt_internal_sbom_handle_qt_entity_license_expression(${target} ${sbom_add_target_args}
             OUT_VAR qt_entity_license_expression)
         if(qt_entity_license_expression)
@@ -859,7 +897,7 @@ function(_qt_internal_sbom_add_target target)
 
     if(license_expression AND
             arg___QT_INTERNAL_HANDLE_QT_ENTITY_TYPE_LICENSE)
-        _qt_internal_sbom_forward_sbom_add_target_options(sbom_add_target_args)
+        _qt_internal_sbom_forward_sbom_add_target_options_modified(sbom_add_target_args)
         _qt_internal_sbom_handle_qt_entity_license_declared_expression(${target}
             ${sbom_add_target_args}
             LICENSE_CONCLUDED_EXPRESSION "${license_expression}"
@@ -877,7 +915,7 @@ function(_qt_internal_sbom_add_target target)
     endif()
 
     if(arg___QT_INTERNAL_HANDLE_QT_ENTITY_TYPE_COPYRIGHTS)
-        _qt_internal_sbom_forward_sbom_add_target_options(sbom_add_target_args)
+        _qt_internal_sbom_forward_sbom_add_target_options_modified(sbom_add_target_args)
         _qt_internal_sbom_handle_qt_entity_copyrights(${target} ${sbom_add_target_args}
             OUT_VAR qt_copyrights)
         if(qt_copyrights)
@@ -901,7 +939,7 @@ function(_qt_internal_sbom_add_target target)
     endif()
 
     if(arg___QT_INTERNAL_HANDLE_QT_ENTITY_TYPE_PACKAGE_VERSION)
-        _qt_internal_sbom_forward_sbom_add_target_options(sbom_add_target_args)
+        _qt_internal_sbom_forward_sbom_add_target_options_modified(sbom_add_target_args)
         _qt_internal_sbom_handle_qt_entity_package_version(${target} ${sbom_add_target_args}
             OUT_VAR qt_entity_package_version)
         if(qt_entity_package_version)
@@ -919,7 +957,7 @@ function(_qt_internal_sbom_add_target target)
     endif()
 
     if(arg___QT_INTERNAL_HANDLE_QT_ENTITY_TYPE_SUPPLIER)
-        _qt_internal_sbom_forward_sbom_add_target_options(sbom_add_target_args)
+        _qt_internal_sbom_forward_sbom_add_target_options_modified(sbom_add_target_args)
         _qt_internal_sbom_handle_qt_entity_supplier(${target} ${sbom_add_target_args}
             OUT_VAR qt_entity_supplier)
         if(qt_entity_supplier)
@@ -937,7 +975,7 @@ function(_qt_internal_sbom_add_target target)
     endif()
 
     if(arg___QT_INTERNAL_HANDLE_QT_ENTITY_TYPE_DOWNLOAD_LOCATION)
-        _qt_internal_sbom_forward_sbom_add_target_options(sbom_add_target_args)
+        _qt_internal_sbom_forward_sbom_add_target_options_modified(sbom_add_target_args)
         _qt_internal_sbom_handle_qt_entity_download_location(${target} ${sbom_add_target_args}
             OUT_VAR qt_entity_download_location)
         if(qt_entity_download_location)
@@ -953,7 +991,7 @@ function(_qt_internal_sbom_add_target target)
         endif()
     endif()
 
-    if(arg_TYPE STREQUAL "SYSTEM_LIBRARY")
+    if(sbom_entity_type STREQUAL "SYSTEM_LIBRARY")
         # Try to get package url that was set using CMake's set_package_properties function.
         # Relies on querying the internal global property name that CMake sets in its
         # implementation.
@@ -972,7 +1010,7 @@ function(_qt_internal_sbom_add_target target)
         list(APPEND project_package_options DOWNLOAD_LOCATION "${download_location}")
     endif()
 
-    _qt_internal_sbom_get_package_purpose("${arg_TYPE}" package_purpose)
+    _qt_internal_sbom_get_package_purpose("${sbom_entity_type}" package_purpose)
     list(APPEND project_package_options PURPOSE "${package_purpose}")
 
     set(cpe_values "")
@@ -1003,7 +1041,7 @@ function(_qt_internal_sbom_add_target target)
     endif()
 
     if(arg___QT_INTERNAL_HANDLE_QT_ENTITY_TYPE_CPE)
-        _qt_internal_sbom_forward_sbom_add_target_options(sbom_add_target_args)
+        _qt_internal_sbom_forward_sbom_add_target_options_modified(sbom_add_target_args)
         _qt_internal_sbom_handle_qt_entity_cpe(${target} ${sbom_add_target_args}
             CPE "${cpe_values}"
             OUT_VAR qt_cpe_list)
@@ -1017,7 +1055,10 @@ function(_qt_internal_sbom_add_target target)
     endif()
 
     # Assemble arguments to forward to the function that handles purl options.
-    set(purl_args "")
+    set(purl_args
+        SBOM_ENTITY_TYPE ${sbom_entity_type}
+    )
+
     _qt_internal_get_sbom_purl_add_target_options(purl_opt_args purl_single_args purl_multi_args)
     _qt_internal_forward_function_args(
         FORWARD_APPEND
@@ -1027,7 +1068,6 @@ function(_qt_internal_sbom_add_target target)
             ${purl_opt_args}
         FORWARD_SINGLE
             ${purl_single_args}
-            TYPE
         FORWARD_MULTI
             ${purl_multi_args}
     )
@@ -1168,7 +1208,7 @@ function(_qt_internal_sbom_add_target target)
         ${no_install_option}
         ${framework_option}
         ${install_prefix_option}
-        TYPE "${arg_TYPE}"
+        SBOM_ENTITY_TYPE "${sbom_entity_type}"
         ${target_binary_multi_config_args}
         SPDX_ID "${package_spdx_id}"
         ${copyrights_option}
@@ -1178,7 +1218,7 @@ function(_qt_internal_sbom_add_target target)
     _qt_internal_sbom_handle_target_custom_files("${target}"
         ${no_install_option}
         ${install_prefix_option}
-        PACKAGE_TYPE "${arg_TYPE}"
+        PACKAGE_TYPE "${sbom_entity_type}"
         PACKAGE_SPDX_ID "${package_spdx_id}"
         ${copyrights_option}
         ${license_option}
@@ -1196,10 +1236,7 @@ function(_qt_internal_add_sbom target)
     set(opt_args
         IMMEDIATE_FINALIZATION
     )
-    set(single_args
-        TYPE
-        FRIENDLY_PACKAGE_NAME
-    )
+    set(single_args "")
     set(multi_args "")
     cmake_parse_arguments(PARSE_ARGV 1 arg "${opt_args}" "${single_args}" "${multi_args}")
     # No validation on purpose, the other options will be validated later.
@@ -1230,9 +1267,7 @@ function(_qt_internal_create_sbom_target target)
     endif()
 
     set(opt_args "")
-    set(single_args
-        TYPE
-    )
+    set(single_args "")
     set(multi_args "")
     cmake_parse_arguments(PARSE_ARGV 1 arg "${opt_args}" "${single_args}" "${multi_args}")
     # No validation on purpose, the other options will be validated later.
@@ -1246,10 +1281,6 @@ function(_qt_internal_create_sbom_target target)
         _qt_sbom_is_custom_sbom_target "TRUE"
         IMPORTED_GLOBAL TRUE
     )
-
-    if(NOT arg_TYPE)
-        message(FATAL_ERROR "No SBOM TYPE option was provided for target: ${target}")
-    endif()
 endfunction()
 
 # Helper to add additional sbom information for an existing target.
@@ -1271,7 +1302,9 @@ function(_qt_internal_extend_sbom target)
         IMMEDIATE_FINALIZATION
     )
     set(single_args
-        TYPE
+        TYPE # deprecated
+        SBOM_ENTITY_TYPE
+        DEFAULT_SBOM_ENTITY_TYPE
         FRIENDLY_PACKAGE_NAME
     )
     set(multi_args "")
@@ -1286,9 +1319,11 @@ function(_qt_internal_extend_sbom target)
         IMMEDIATE_FINALIZATION
     )
 
+    _qt_internal_map_sbom_entity_type(sbom_entity_type ${ARGN})
+
     # Make sure a spdx id is recorded for the target right now, so it is "known" when handling
     # relationships for other targets, even if the target was not yet finalized.
-    if(arg_TYPE)
+    if(sbom_entity_type)
         # Friendly package name is allowed to be empty.
         set(package_name_option "")
         if(arg_FRIENDLY_PACKAGE_NAME)
@@ -1296,7 +1331,7 @@ function(_qt_internal_extend_sbom target)
         endif()
 
         _qt_internal_sbom_record_target_spdx_id(${target}
-            TYPE "${arg_TYPE}"
+            SBOM_ENTITY_TYPE "${sbom_entity_type}"
             ${package_name_option}
         )
     endif()
@@ -1519,7 +1554,7 @@ function(_qt_internal_sbom_record_target_spdx_id target)
     set(opt_args "")
     set(single_args
         PACKAGE_NAME
-        TYPE
+        SBOM_ENTITY_TYPE
         OUT_VAR
     )
     set(multi_args "")
@@ -1543,12 +1578,12 @@ function(_qt_internal_sbom_record_target_spdx_id target)
     endif()
 
     _qt_internal_sbom_generate_target_package_spdx_id(package_spdx_id
-        TYPE "${arg_TYPE}"
+        SBOM_ENTITY_TYPE "${arg_SBOM_ENTITY_TYPE}"
         PACKAGE_NAME "${package_name_for_spdx_id}"
     )
     _qt_internal_sbom_save_spdx_id_for_target("${target}" "${package_spdx_id}")
 
-    _qt_internal_sbom_is_qt_entity_type("${arg_TYPE}" is_qt_entity_type)
+    _qt_internal_sbom_is_qt_entity_type("${arg_SBOM_ENTITY_TYPE}" is_qt_entity_type)
     _qt_internal_sbom_save_spdx_id_for_qt_entity_type(
         "${target}" "${is_qt_entity_type}" "${package_spdx_id}")
 
@@ -1562,7 +1597,7 @@ function(_qt_internal_sbom_generate_target_package_spdx_id out_var)
     set(opt_args "")
     set(single_args
         PACKAGE_NAME
-        TYPE
+        SBOM_ENTITY_TYPE
     )
     set(multi_args "")
     cmake_parse_arguments(PARSE_ARGV 1 arg "${opt_args}" "${single_args}" "${multi_args}")
@@ -1571,12 +1606,12 @@ function(_qt_internal_sbom_generate_target_package_spdx_id out_var)
     if(NOT arg_PACKAGE_NAME)
         message(FATAL_ERROR "PACKAGE_NAME must be set")
     endif()
-    if(NOT arg_TYPE)
-        message(FATAL_ERROR "TYPE must be set")
+    if(NOT arg_SBOM_ENTITY_TYPE)
+        message(FATAL_ERROR "SBOM_ENTITY_TYPE must be set")
     endif()
 
     _qt_internal_sbom_get_root_project_name_for_spdx_id(repo_project_name_spdx_id)
-    _qt_internal_sbom_get_package_infix("${arg_TYPE}" package_infix)
+    _qt_internal_sbom_get_package_infix("${arg_SBOM_ENTITY_TYPE}" package_infix)
 
     _qt_internal_sbom_get_sanitized_spdx_id(spdx_id
         "SPDXRef-${repo_project_name_spdx_id}-${package_infix}-${arg_PACKAGE_NAME}")
diff --git a/cmake/QtPublicSbomQtEntityHelpers.cmake b/cmake/QtPublicSbomQtEntityHelpers.cmake
index 656f04ca1d8..dfe3bf2be0a 100644
--- a/cmake/QtPublicSbomQtEntityHelpers.cmake
+++ b/cmake/QtPublicSbomQtEntityHelpers.cmake
@@ -4,7 +4,7 @@
 # Helper macro to prepare forwarding all set sbom options to some other function.
 # Expects the options names to be set in the parent scope by calling
 # _qt_internal_get_sbom_add_target_options(opt_args single_args multi_args)
-macro(_qt_internal_sbom_forward_sbom_add_target_options args_var_name)
+macro(_qt_internal_sbom_forward_sbom_add_target_options_modified args_var_name)
     if(NOT opt_args)
         message(FATAL_ERROR
             "Expected opt_args to be set by _qt_internal_get_sbom_add_target_options")
@@ -17,16 +17,35 @@ macro(_qt_internal_sbom_forward_sbom_add_target_options args_var_name)
         message(FATAL_ERROR
             "Expected multi_args to be set by _qt_internal_get_sbom_add_target_options")
     endif()
+
+    if(NOT sbom_entity_type)
+        message(FATAL_ERROR
+            "Expected sbom_entity_type to be set by the parent scope")
+    endif()
+
+    set(modified_opt_args ${opt_args})
+    set(modified_single_args ${single_args})
+    set(modified_multi_args ${multi_args})
+
+    # Don't forward the sbom entity type values, as they are handled manually.
+    list(REMOVE_ITEM modified_single_args
+        TYPE
+        SBOM_ENTITY_TYPE
+        DEFAULT_SBOM_ENTITY_TYPE
+    )
+
     _qt_internal_forward_function_args(
         FORWARD_PREFIX arg
         FORWARD_OUT_VAR ${args_var_name}
         FORWARD_OPTIONS
-            ${opt_args}
+            ${modified_opt_args}
         FORWARD_SINGLE
-            ${single_args}
+            ${modified_single_args}
         FORWARD_MULTI
-            ${multi_args}
+            ${modified_multi_args}
     )
+
+    list(APPEND ${args_var_name} SBOM_ENTITY_TYPE "${sbom_entity_type}")
 endmacro()
 
 # Helper function to add a default supplier for a qt entity type.
@@ -36,8 +55,9 @@ function(_qt_internal_sbom_handle_qt_entity_supplier target)
     cmake_parse_arguments(PARSE_ARGV 1 arg "${opt_args}" "${single_args}" "${multi_args}")
     _qt_internal_validate_all_args_are_parsed(arg)
 
-    _qt_internal_sbom_is_qt_entity_type("${arg_TYPE}" is_qt_entity_type)
-    _qt_internal_sbom_is_qt_3rd_party_entity_type("${arg_TYPE}" is_qt_3rd_party_entity_type)
+    _qt_internal_sbom_is_qt_entity_type("${arg_SBOM_ENTITY_TYPE}" is_qt_entity_type)
+    _qt_internal_sbom_is_qt_3rd_party_entity_type("${arg_SBOM_ENTITY_TYPE}"
+        is_qt_3rd_party_entity_type)
 
     set(supplier "")
     if(NOT arg_SUPPLIER
@@ -58,7 +78,7 @@ function(_qt_internal_sbom_handle_qt_entity_package_version target)
     cmake_parse_arguments(PARSE_ARGV 1 arg "${opt_args}" "${single_args}" "${multi_args}")
     _qt_internal_validate_all_args_are_parsed(arg)
 
-    _qt_internal_sbom_is_qt_entity_type("${arg_TYPE}" is_qt_entity_type)
+    _qt_internal_sbom_is_qt_entity_type("${arg_SBOM_ENTITY_TYPE}" is_qt_entity_type)
 
     set(package_version "")
     if(NOT arg_PACKAGE_VERSION
@@ -79,7 +99,7 @@ function(_qt_internal_sbom_handle_qt_entity_download_location target)
     cmake_parse_arguments(PARSE_ARGV 1 arg "${opt_args}" "${single_args}" "${multi_args}")
     _qt_internal_validate_all_args_are_parsed(arg)
 
-    _qt_internal_sbom_is_qt_entity_type("${arg_TYPE}" is_qt_entity_type)
+    _qt_internal_sbom_is_qt_entity_type("${arg_SBOM_ENTITY_TYPE}" is_qt_entity_type)
 
     set(download_location "")
     if(NOT arg_DOWNLOAD_LOCATION AND is_qt_entity_type)
@@ -98,13 +118,13 @@ function(_qt_internal_sbom_handle_qt_entity_license_expression target)
     cmake_parse_arguments(PARSE_ARGV 1 arg "${opt_args}" "${single_args}" "${multi_args}")
     _qt_internal_validate_all_args_are_parsed(arg)
 
-    _qt_internal_sbom_is_qt_entity_type("${arg_TYPE}" is_qt_entity_type)
+    _qt_internal_sbom_is_qt_entity_type("${arg_SBOM_ENTITY_TYPE}" is_qt_entity_type)
 
     set(license_expression "")
 
     # For Qt entities, we have some special handling.
     if(is_qt_entity_type AND NOT arg_NO_DEFAULT_QT_LICENSE AND NOT arg_QT_LICENSE_ID)
-        if(arg_TYPE STREQUAL "QT_TOOL" OR arg_TYPE STREQUAL "QT_APP")
+        if(arg_SBOM_ENTITY_TYPE STREQUAL "QT_TOOL" OR arg_SBOM_ENTITY_TYPE STREQUAL "QT_APP")
             if(QT_SBOM_DEFAULT_QT_LICENSE_ID_EXECUTABLES
                     AND NOT arg_NO_DEFAULT_QT_LICENSE_ID_EXECUTABLES)
                 # A repo might contain only the "gpl3" license variant as the default for all
@@ -161,7 +181,7 @@ function(_qt_internal_sbom_handle_qt_entity_license_declared_expression target)
     cmake_parse_arguments(PARSE_ARGV 1 arg "${opt_args}" "${single_args}" "${multi_args}")
     _qt_internal_validate_all_args_are_parsed(arg)
 
-    _qt_internal_sbom_is_qt_entity_type("${arg_TYPE}" is_qt_entity_type)
+    _qt_internal_sbom_is_qt_entity_type("${arg_SBOM_ENTITY_TYPE}" is_qt_entity_type)
 
     set(license_expression "")
 
@@ -189,7 +209,7 @@ function(_qt_internal_sbom_handle_qt_entity_copyrights target)
     cmake_parse_arguments(PARSE_ARGV 1 arg "${opt_args}" "${single_args}" "${multi_args}")
     _qt_internal_validate_all_args_are_parsed(arg)
 
-    _qt_internal_sbom_is_qt_entity_type("${arg_TYPE}" is_qt_entity_type)
+    _qt_internal_sbom_is_qt_entity_type("${arg_SBOM_ENTITY_TYPE}" is_qt_entity_type)
 
     set(qt_default_copyright "")
     if(is_qt_entity_type AND NOT arg_NO_DEFAULT_QT_COPYRIGHTS)
@@ -208,8 +228,9 @@ function(_qt_internal_sbom_handle_qt_entity_cpe target)
     cmake_parse_arguments(PARSE_ARGV 1 arg "${opt_args}" "${single_args}" "${multi_args}")
     _qt_internal_validate_all_args_are_parsed(arg)
 
-    _qt_internal_sbom_is_qt_entity_type("${arg_TYPE}" is_qt_entity_type)
-    _qt_internal_sbom_is_qt_3rd_party_entity_type("${arg_TYPE}" is_qt_3rd_party_entity_type)
+    _qt_internal_sbom_is_qt_entity_type("${arg_SBOM_ENTITY_TYPE}" is_qt_entity_type)
+    _qt_internal_sbom_is_qt_3rd_party_entity_type("${arg_SBOM_ENTITY_TYPE}"
+        is_qt_3rd_party_entity_type)
 
     set(cpe_list "")
 
@@ -380,8 +401,9 @@ function(_qt_internal_sbom_handle_qt_entity_purl_entries)
 
     set(purl_ids "")
 
-    _qt_internal_sbom_is_qt_entity_type("${arg_TYPE}" is_qt_entity_type)
-    _qt_internal_sbom_is_qt_3rd_party_entity_type("${arg_TYPE}" is_qt_3rd_party_entity_type)
+    _qt_internal_sbom_is_qt_entity_type("${arg_SBOM_ENTITY_TYPE}" is_qt_entity_type)
+    _qt_internal_sbom_is_qt_3rd_party_entity_type("${arg_SBOM_ENTITY_TYPE}"
+        is_qt_3rd_party_entity_type)
 
     if(is_qt_entity_type OR is_qt_3rd_party_entity_type)
         _qt_internal_sbom_get_qt_entity_default_purl_ids(purl_ids)
diff --git a/cmake/QtPublicSbomSystemDepHelpers.cmake b/cmake/QtPublicSbomSystemDepHelpers.cmake
index f2bee2a73cd..24967a42801 100644
--- a/cmake/QtPublicSbomSystemDepHelpers.cmake
+++ b/cmake/QtPublicSbomSystemDepHelpers.cmake
@@ -11,7 +11,8 @@ function(_qt_internal_sbom_record_system_library_usage target)
 
     set(opt_args "")
     set(single_args
-        TYPE
+        TYPE # deprecated
+        SBOM_ENTITY_TYPE
         PACKAGE_VERSION
         FRIENDLY_PACKAGE_NAME
     )
@@ -19,8 +20,10 @@ function(_qt_internal_sbom_record_system_library_usage target)
     cmake_parse_arguments(PARSE_ARGV 1 arg "${opt_args}" "${single_args}" "${multi_args}")
     _qt_internal_validate_all_args_are_parsed(arg)
 
-    if(NOT arg_TYPE)
-        message(FATAL_ERROR "TYPE must be set")
+    _qt_internal_map_sbom_entity_type(sbom_entity_type ${ARGN})
+
+    if(NOT sbom_entity_type)
+        message(FATAL_ERROR "SBOM_ENTITY_TYPE is empty for target '${target}', but it must be set")
     endif()
 
     # A package might be looked up more than once, make sure to record it once.
@@ -40,7 +43,7 @@ function(_qt_internal_sbom_record_system_library_usage target)
     # has started, e.g. during _qt_internal_find_third_party_dependencies.
     set(spdx_options
         ${target}
-        TYPE "${arg_TYPE}"
+        SBOM_ENTITY_TYPE "${sbom_entity_type}"
         PACKAGE_NAME "${arg_FRIENDLY_PACKAGE_NAME}"
     )
 
diff --git a/cmake/QtTargetHelpers.cmake b/cmake/QtTargetHelpers.cmake
index 2c8d79b89da..37363ed0d26 100644
--- a/cmake/QtTargetHelpers.cmake
+++ b/cmake/QtTargetHelpers.cmake
@@ -1839,7 +1839,7 @@ function(qt_internal_add_platform_internal_target target)
     qt_internal_mark_as_internal_library("${target}")
 
     qt_internal_add_sbom("${target}"
-        TYPE QT_MODULE
+        SBOM_ENTITY_TYPE QT_MODULE
     )
 endfunction()
 
diff --git a/cmake/QtToolHelpers.cmake b/cmake/QtToolHelpers.cmake
index a7e1842c04f..13d38967cf0 100644
--- a/cmake/QtToolHelpers.cmake
+++ b/cmake/QtToolHelpers.cmake
@@ -211,7 +211,7 @@ function(qt_internal_add_tool target_name)
 
     if(QT_GENERATE_SBOM)
         set(sbom_args "")
-        list(APPEND sbom_args TYPE QT_TOOL)
+        list(APPEND sbom_args DEFAULT_SBOM_ENTITY_TYPE QT_TOOL)
     endif()
 
     if(NOT arg_NO_INSTALL AND arg_TOOLS_TARGET)
diff --git a/qmake/CMakeLists.txt b/qmake/CMakeLists.txt
index 012d22d909e..1b1083f84df 100644
--- a/qmake/CMakeLists.txt
+++ b/qmake/CMakeLists.txt
@@ -17,7 +17,7 @@ if(QT_WILL_BUILD_TOOLS)
         qmakelibraryinfo.cpp qmakelibraryinfo.h
     )
     qt_internal_add_sbom(QtLibraryInfo
-        TYPE QT_MODULE
+        SBOM_ENTITY_TYPE QT_MODULE
         NO_INSTALL
     )
 
diff --git a/src/tools/bootstrap/CMakeLists.txt b/src/tools/bootstrap/CMakeLists.txt
index 399e2419104..6fcfc4855b1 100644
--- a/src/tools/bootstrap/CMakeLists.txt
+++ b/src/tools/bootstrap/CMakeLists.txt
@@ -9,7 +9,7 @@
 # libraries.
 qt_add_library(Bootstrap STATIC)
 qt_internal_add_sbom(Bootstrap
-    TYPE QT_MODULE
+    SBOM_ENTITY_TYPE QT_MODULE
     NO_INSTALL
 )
 
diff --git a/tests/auto/cmake/RunCMake/Sbom/common_targets.cmake b/tests/auto/cmake/RunCMake/Sbom/common_targets.cmake
index 0c6b35596c9..bde4ab96ee1 100644
--- a/tests/auto/cmake/RunCMake/Sbom/common_targets.cmake
+++ b/tests/auto/cmake/RunCMake/Sbom/common_targets.cmake
@@ -59,3 +59,27 @@ endif()
 
 add_subdirectory(custom_files)
 
+# Exercise various sbom entity options.
+foreach(idx RANGE 2)
+    if(idx EQUAL "0")
+        set(option_name "TYPE")
+    elseif(idx EQUAL "1")
+        set(option_name "SBOM_ENTITY_TYPE")
+    else()
+        set(option_name "DEFAULT_SBOM_ENTITY_TYPE")
+    endif()
+    set(target lib_various_types_${idx})
+    add_library(${target} STATIC)
+    target_sources(${target} PRIVATE sources/core_helper.cpp)
+    install(TARGETS ${target}
+        RUNTIME DESTINATION bin
+        ARCHIVE DESTINATION lib
+        LIBRARY DESTINATION lib
+    )
+    _qt_internal_add_sbom(${target}
+        ${option_name} "LIBRARY"
+        RUNTIME_PATH bin
+        ARCHIVE_PATH lib
+        LIBRARY_PATH lib
+    )
+endforeach()
-- 
2.51.2

